packages/firestore/src/platform_node/insecure_grpc_connection.ts

```diff
@@ -0,0 +1,283 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import * as grpc from 'grpc';
+
+import firebase from '@firebase/app';
+const SDK_VERSION = firebase.SDK_VERSION;
+
+const grpcVersion = require('grpc/package.json').version;
+
+import { Token } from '../api/credentials';
+import { DatabaseInfo } from '../core/database_info';
+import { Connection, Stream } from '../remote/connection';
+import { StreamBridge } from '../remote/stream_bridge';
+import { mapCodeFromRpcCode } from '../remote/rpc_error';
+import { assert } from '../util/assert';
+import { FirestoreError } from '../util/error';
+import * as log from '../util/log';
+import { NodeCallback, nodePromise } from '../util/node_api';
+import { Deferred } from '../util/promise';
+
+const LOG_TAG = 'Insecure Connection';
+
+// TODO(b/38203344): The SDK_VERSION is set independently from Firebase because
+// we are doing out-of-band releases. Once we release as part of Firebase, we
+// should use the Firebase version instead.
+const X_GOOG_API_CLIENT_VALUE = `gl-node/${
+  process.versions.node
+} fire/${SDK_VERSION} grpc/${grpcVersion}`;
+
+type DuplexRpc<Req, Resp> = () => grpc.ClientDuplexStream<Req, Resp>;
+type ReadableRpc<Req, Resp> = (req: Req) => grpc.ClientReadableStream<Resp>;
+type UnaryRpc<Req, Resp> = (
+  req: Req,
+  callback: (err?: grpc.ServiceError, resp?: Resp) => void
+) => grpc.ClientUnaryCall;
+
+// The type of these stubs is dynamically generated by the GRPC runtime
+// from the protocol buffer.
+// tslint:disable-next-line:no-any
+type GeneratedGrpcStub = any;
+
+interface CachedStub {
+  stub: GeneratedGrpcStub;
+  token: Token | null;
+}
+
+/**
+ * A Connection implemented by GRPC-Node.
+ */
+export class InsecureGrpcConnection implements Connection {
+  // tslint:disable-next-line:no-any
+  private firestore: any;
+
+  // We cache stubs for the most-recently-used token.
+  private cachedStub: CachedStub | null = null;
+
+  constructor(protos: grpc.GrpcObject, private databaseInfo: DatabaseInfo) {
+    this.firestore = protos['google']['firestore']['v1beta1'];
+  }
+
+  private sameToken(tokenA: Token | null, tokenB: Token | null): boolean {
+    const valueA = tokenA && tokenA.authHeaders['Authorization'];
+    const valueB = tokenB && tokenB.authHeaders['Authorization'];
+    return valueA === valueB;
+  }
+
+  private ensureActiveStub(token: Token | null): GeneratedGrpcStub {
+    if (!this.cachedStub || !this.sameToken(this.cachedStub.token, token)) {
+      log.debug(LOG_TAG, 'Creating Firestore stub.');
+      const credentials = grpc.credentials.createInsecure();
+      this.cachedStub = {
+        stub: new this.firestore.Firestore(this.databaseInfo.host, credentials),
+        token
+      };
+    }
+
+    return this.cachedStub.stub;
+  }
+
+  private getRpcCallable<Req, Resp>(
```

Comment:
There is a huge amount of duplicated code between `grpc_connection` and `insecure_grpc_connection`. This is basically the only method that's actually different. Instead of constructing a secure set of channel credentials with the authentication metadata baked in via call credentials, it uses insecure credentials and directly attaches the metadata to each RPC.